app/packages/commands/src/keys/manager.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright 2017-2025, Voxel51, Inc.
+ */
+
+import { Command } from "../types";
+import { KeySequence, KeyParser } from "./parser";
+import { CommandRegistry, getCommandRegistry } from "../registry";
+
+/**
+ * The order of this enum determines the precedence of the scopes
+ */
+export enum KeyBindingScope {
+  User = "User",
+  Plugin = "Plugin",
+  Core = "Core",
+}
+
+type MatchState = {
+  partial: boolean;
+  full?: Command;
+};
+
+/**
+ * Manages a set of key bindings in scopes.
+ * Each scope must contain unique bindings.  
+ * Scopes may shadow bindings in other scopes with the 
+ * precedence being User, Plugin, Core.
+ * This allows plugins and user re-bindings.
+ * @see KeyParser for documentation on the binding format.
+ */
+export class KeyManager {
```

Comment:
nit: I see that we have a global instance of this and therefore not relevant _except for tests_, but maybe we should add a `cleanup()` equivalent function that removes the event listener.